spark/src/test/scala/ai/chronon/spark/test/JoinFlowTest.scala

```diff
@@ -0,0 +1,123 @@
+      Column("amount", LongType, 1000)
+    )
+    val transactionsTable = s"$namespace.${prefix}_transactions"
+    DataFrameGen
+      .events(spark, transactionSchema, 1200, partitions = 30)
+      .where(col("user").isNotNull)
+      .save(transactionsTable)
+
+    val joinPart: JoinPart = Builders.JoinPart(groupBy = Builders.GroupBy(
+      keyColumns = Seq("user"),
+      sources = Seq(
+        Builders.Source.events(
+          query = Builders.Query(
+            selects = Builders.Selects("amount"),
+            timeColumn = "ts"
+          ),
+          table = transactionsTable
+        )),
+      aggregations = Seq(
+        Builders.Aggregation(operation = Operation.SUM,
+                             inputColumn = "amount",
+                             windows = Seq(new Window(7, TimeUnit.DAYS), new Window(14, TimeUnit.DAYS)))),
+      accuracy = Accuracy.SNAPSHOT,
+      metaData = Builders.MetaData(name = s"join_test.${prefix}_txn", namespace = namespace, team = "chronon")
+    ))
+
+    // bootstrap part
+    val bootstrapSourceTable = s"$namespace.${prefix}_bootstrap_source"
+    spark
+      .table(queryTable)
+      .sample(0.5)
+      .withColumn(s"${joinPart.fullPrefix}_amount_sum_7d", lit(0).cast("long"))
+      .withColumn(s"${joinPart.fullPrefix}_amount_sum_14d", lit(0).cast("long"))
+      .save(bootstrapSourceTable)
+    val bootstrapPart = Builders.BootstrapPart(
+      table = bootstrapSourceTable,
+      keyColumns = Seq("user"),
+      query = Builders.Query(
+        Builders.Selects("user", s"${joinPart.fullPrefix}_amount_sum_7d", s"${joinPart.fullPrefix}_amount_sum_14d"))
+    )
+
+    // derivations
+    val derivations = Seq(
+      Builders.Derivation(
+        name = s"${joinPart.fullPrefix}_amount_sum_7d",
+        expression = s"COALESCE(${joinPart.fullPrefix}_amount_sum_7d, 0)"
+      ),
+      Builders.Derivation(
+        name = s"${joinPart.fullPrefix}_amount_sum_14d",
+        expression = s"COALESCE(${joinPart.fullPrefix}_amount_sum_14d, 0)"
+      )
+    )
+
+    // join
+    val join = Builders.Join(
+      left = querySource,
+      joinParts = Seq(joinPart),
+      bootstrapParts = Seq(bootstrapPart),
+      rowIds = Seq("user"),
+      derivations = derivations,
+      metaData = Builders.MetaData(name = s"unit_test.${prefix}_join", namespace = namespace, team = "chronon")
+    )
+
+    val endDs = tableUtils.partitions(queryTable).max
+    val joinJob = new Join(join, endDs, tableUtils)
+
+    // compute left
+    joinJob.computeLeft()
+    assertTrue(tableUtils.tableExists(join.metaData.bootstrapTable))
+
+    // compute right
+    val joinPartJob = new Join(join, endDs, tableUtils, selectedJoinParts = Some(List(joinPart.fullPrefix)))
+    joinPartJob.computeJoinOpt(useBootstrapForLeft = true)
+    assertTrue(tableUtils.tableExists(join.partOutputTable(joinPart)))
+
+    // compute final
+    joinJob.computeFinal()
+    assertTrue(tableUtils.tableExists(join.metaData.outputTable))
```

Comment:
@donghanz FYI, I added a new UT that closely follows the Flow mode (left => join part => final). Previously this was not covered by UT